File: leetcode/number-of-1-bits.py
# https://leetcode.com/problems/number-of-1-bits/

# Tags: Array - Math

import timeit


# Shift bits and check if the last bit is a 1
# https://stackoverflow.com/a/13522788/2557030
#
# Time complexity: O(log(n))
# Space complexity: O(1)
#
# Runtime: 41 ms, faster than 69.96% of Python3 online submissions for Number of 1 Bits.
# Memory Usage: 13.8 MB, less than 50.25% of Python3 online submissions for Number of 1 Bits.
class CountBits:
    def hammingWeight(self, n: int) -> int:
        ones = 0
        while n:
            # Add one if the last bit is even
            if n % 2 == 1:
                ones += 1
            # Shift bits to the right
            n >>= 1
        return ones


# Use the bit_count method added in python 3.10.
#
# Runtime: 37 ms, faster than 81.43% of Python3 online submissions for Number of 1 Bits.
# Memory Usage: 14 MB, less than 7.72% of Python3 online submissions for Number of 1 Bits.
class BuiltIn:
    def hammingWeight(self, n: int) -> int:
        # Needs python >= 3.10
        return n.bit_count()


def test():
    executors = [CountBits, BuiltIn]
    tests = [
        [0, 0],
        [11, 3],
        [128, 1],
        [4294967293, 31],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(int(float("1"))):
            for i, t in enumerate(tests):
                sol = executor()
                result = sol.hammingWeight(t[0])
                exp = t[1]
                assert result == exp, f"\033[93m» {t[0]} <> {exp}\033[91m for test {i} using \033[1m{executor.__name__}"
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        res = "{0:20}{1:10}{2:10}".format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/number-of-1-bits.rs
// 191. Number of 1 Bits
// 🟢 Easy
//
// https://leetcode.com/problems/number-of-1-bits/
//
// Tags: Divide and Conquer - Bit Manipulation

struct Solution;
impl Solution {
    #![allow(dead_code)]
    /// And the number with 1 to check if the least significant bit is set, then shift bits by one
    /// to move into the next one.
    ///
    /// Time complexity: O(log2(n)) - At each iteration we divide the number by 2.
    /// Space complexity: O(1) - We store one i32 value.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 1.97 MB Beats 91.94%
    pub fn hamming_weight(n: u32) -> i32 {
        let mut count_ones = 0;
        let mut n = n;
        while n > 0 {
            count_ones += n & 1;
            n >>= 1;
        }
        count_ones as i32
    }

    /// Add n to n-1, that removes the least significant set bit, this solution will iterate only m
    /// times, where m is the number of '1' bits, while the previous one also does one iteration
    /// per each '0' bit.
    ///
    /// Time complexity: O(m) - Where m is the number of '1' bits and it can be, at most log2(n)
    /// Space complexity: O(1) - We store one i32 value.
    ///
    /// Runtime 0 ms Beats 100%
    /// Memory 1.99 MB Beats 91.74%
    pub fn hamming_weight_2(n: u32) -> i32 {
        let mut count_ones = 0;
        let mut n = n;
        while n > 0 {
            n &= n - 1;
            count_ones += 1;
        }
        count_ones as i32
    }
}

// Tests.
fn main() {
    let tests = [
        // 00000000000000000000000000001011
        (11, 3),
        // 00000000000000000000000010000000
        (128, 1),
        // 11111111111111111111111111111101
        (4294967293, 31),
    ];
    println!("\n\x1b[92m» Running {} tests...\x1b[0m", tests.len());
    let mut success = 0;
    for (i, t) in tests.iter().enumerate() {
        let res = Solution::hamming_weight_2(t.0.clone());
        if res == t.1 {
            success += 1;
            println!("\x1b[92m✔\x1b[95m Test {} passed!\x1b[0m", i);
        } else {
            println!(
                "\x1b[31mx\x1b[95m Test {} failed expected: {:?} but got {}!!\x1b[0m",
                i, t.1, res
            );
        }
    }
    println!("");
    if success == tests.len() {
        println!("\x1b[30;42m✔ All tests passed!\x1b[0m")
    } else if success == 0 {
        println!("\x1b[31mx \x1b[41;37mAll tests failed!\x1b[0m")
    } else {
        println!(
            "\x1b[31mx\x1b[95m {} tests failed!\x1b[0m",
            tests.len() - success
        )
    }
}
